README.md
# (MODX)EvolutionCMS.plugins.ManagerManager.mm_ddMaxLength

Widget for ManagerManager plugin allowing number limitation of chars inputing in document fields (including TVs).


## Requires

* PHP >= 5.4
* [(MODX)EvolutionCMS.plugins.ManagerManager](https://code.divandesign.biz/modx/managermanager) >= 0.7


## Documentation


### Installation

To install you must unzip the archive to `/assets/plungins/managermanager/widgets/`.


You may also read this documentation:
* [(MODX)EvolutionCMS.plugins.ManagerManager](https://code.divandesign.biz/modx/managermanager).
* [(MODX)EvolutionCMS.modules.ddMMEditor](https://code.divandesign.biz/modx/ddmmeditor).


### Parameters description

* `$params`
	* Desctription: Parameters, the pass-by-name style is used.
	* Valid values:
		* `stdClass`
		* `arrayAssociative`
	* **Required**
	
* `$params->fields`
	* Desctription: The names of TVs for which the widget is applied.  
	* Valid values: `stringCommaSeparated`
	* **Required**
	
* `$params->length`
	* Desctription: Maximum number of inputing chars.
	* Valid values: `integer`
	* Default value: `150`
	
* `$params->allowTypingOverLimit`
	* Desctription: Is typing over limit allowed?  
		In this case users can enter any length text but they can't save document if the limit is over.
	* Valid values: `boolean`
	* Default value: `true`
	
* `$params->roles`
	* Desctription: The CMS user roles that the widget is applied to.
	* Valid values:
		* `stringCommaSeparated`
		* `''` — when this parameter is empty then widget is applied to the all roles
	* Default value: `''`
	
* `$params->roles[i]`
	* Desctription: CMS user role.
	* Valid values: `integer`
	* **Required**
	
* `$params->templates`
	* Desctription: Document templates IDs for which the widget is applying to.
	* Valid values:
		* `stringCommaSeparated`
		* `''` — empty value means the widget is applying to all templates
	* Default value: `''`
	
* `$params->templates[i]`
	* Desctription: Templates ID.
	* Valid values: `integer`
	* **Required**


### CMS events

* `OnDocFormPrerender`
* `OnDocFormRender`


### Examples


#### Limit the number of inputting chars to 140 in `pagetitle` and `menutitle`

```
mm_ddMaxLength([
	'fields' => 'pagetitle,menutitle',
	'length' => 140
]);
```


## Links

* [Home page](https://code.divandesign.biz/modx/mm_ddmaxlength)
* [Telegram chat](https://t.me/dd_code)
* [Packagist](https://packagist.org/packages/dd/evolutioncms-plugins-managermanager-mm_ddmaxlength)


<link rel="stylesheet" type="text/css" href="https://DivanDesign.ru/assets/files/ddMarkdown.css" />

CHANGELOG.md
# (MODX)EvolutionCMS.plugins.ManagerManager.mm_ddMaxLength changelog


## Version 1.2.1 (2016-12-06)
* \* Parameters → `$params->length`: Is cast to integer.


## Version 1.2 (2016-11-15)
* \* Attention! PHP >= 5.4 is required.
* \* Attention! (MODX)EvolutionCMS.plugins.ManagerManager >= 0.7 is required.
* \* Refactoring:
	* \* The widget now using named parameters (with backward compatibility).
	* \* `jQuery.ddMM.fields` is used istead of manually DOM finding.
	* \* File name refactoring.


## Version 1.1.1 (2013-12-10)
* \* The calls of the functions `includeJs` and `includeCss` were replaced by `includeJsCss`.
* \* The including of required JS and CSS files is currently being realized while occuring the event `OnDocFormPrerender`. The files are included as HTML.
* \* The JS code became a stand-alone file and was partially revised. It is convenient because of php code cleanness and it shortens the amount of code of a document edit frame.
* \- The `jQuery.ddTools` library is no longer included here because it is in `ManagerManager`.


## Version 1.1 (2013-10-02)
* \* Attention! (MODX)EvolutionCMS.plugins.ManagerManager >= 0.6 is required.
* \+ The widget now can also be applied to document fields.
* \* Parameters → `fields`: Was renamed from `tvs`.
* \* The file `jquery.ddmaxlength-1.0.min.js` has been deleted because the ManagerManager plugin already contains the `jQuery.ddTools` library including the `jQuery.fn.ddMaxLength` plugin.
* \* Misprints in classnames have been corrected.


## Version 1.0.1 (2012-01-13)
* \* Added checking of the `OnDocFormRender` event.


<link rel="stylesheet" type="text/css" href="https://DivanDesign.ru/assets/files/ddMarkdown.css" />
<style>ul{list-style:none;}</style>

ddmaxlength.php
<?php
/**
 * mm_ddMaxLength
 * @version 1.2.1 (2016-12-06)
 * 
 * @see README.md
 * 
 * @link https://code.divandesign.biz/modx/mm_ddmaxlength
 * 
 * @copyright 2012–2016 DD Group {@link https://DivanDesign.biz }
 */

function mm_ddMaxLength($params){
	//For backward compatibility
	if (
		!is_array($params) &&
		!is_object($params)
	){
		//Convert ordered list of params to named
		$params = \ddTools::orderedParamsToNamed([
			'paramsList' => func_get_args(),
			'compliance' => [
				'fields',
				'roles',
				'templates',
				'length'
			]
		]);
	}
	
	//Defaults
	$params = \DDTools\ObjectTools::extend([
		'objects' => [
			(object) [
				'fields' => '',
				'length' => 150,
				'allowTypingOverLimit' => true,
				'roles' => '',
				'templates' => ''
			],
			$params
		]
	]);
	
	if (
		!useThisRule(
			$params->roles,
			$params->templates
		)
	){
		return;
	}
	
	global $modx;
	$e = &$modx->Event;
	
	$output = '';
	
	if ($e->name == 'OnDocFormPrerender'){
		$widgetDir =
			$modx->getConfig('site_url') .
			'assets/plugins/managermanager/widgets/ddmaxlength/'
		;
		
		$output .= includeJsCss(
			(
				$widgetDir .
				'ddmaxlength.css'
			),
			'html'
		);
		$output .= includeJsCss(
			(
				$widgetDir .
				'jQuery.ddMM.mm_ddMaxLength.js'
			),
			'html',
			'jQuery.ddMM.mm_ddMaxLength',
			'1.0.1'
		);
		
		$e->output($output);
	}else if ($e->name == 'OnDocFormRender'){
		$params->fields = getTplMatchedFields(
			$params->fields,
			'text,textarea'
		);
		
		if ($params->fields == false){
			return;
		}
		
		$output .=
			'//---------- mm_ddMaxLength :: Begin -----' .
			PHP_EOL
		;
		
		$output .=
'
$j.ddMM
	.getFieldElems({
		fields: ' .
			\DDTools\ObjectTools::convertType([
				'object' => $params->fields,
				'type' => 'stringJsonArray'
			]) .
		'
	}).
	.addClass("ddMaxLengthField")
	.each(function(){
		$j(this)
			.parent()
			.append("<div class=\"ddMaxLengthCount\"><span></span></div>")
		;
	})
	.ddMaxLength({
		max: ' . intval($params->length) . ',
		canWriteError: ' . intval($params->allowTypingOverLimit) . ',
		containerSelector: "div.ddMaxLengthCount span",
		warningClass: "maxLengthWarning"
	})
;
'
		;
		
		$output .=
			'//---------- mm_ddMaxLength :: End -----' .
			PHP_EOL
		;
		
		$e->output($output);
	}
}
?>